b2/sync.py

```diff
@@ -421,39 +649,115 @@ def zip_folders(folder_a, folder_b):
+        )  # yapf: disable
+    else:
+        return B2DownloadAction(
+            source_file.name,
+            source_folder.make_full_path(source_file.name),
+            source_file.latest_version().id_,
+            dest_folder.make_full_path(source_file.name),
+            source_mod_time,
+            source_file.latest_version().size
+        )  # yapf: disable
+
+
 def make_file_sync_actions(
-    sync_type, source_file, dest_file, source_folder, dest_folder, history_days
+    sync_type, source_file, dest_file, source_folder, dest_folder, args, now_millis
 ):
     """
     Yields the sequence of actions needed to sync the two files
     """
+    # Get the modification time of the latest version of the source file
     source_mod_time = 0
     if source_file is not None:
         source_mod_time = source_file.latest_version().mod_time
+
+    # Get the modification time of the latest version of the destination file
     dest_mod_time = 0
     if dest_file is not None:
         dest_mod_time = dest_file.latest_version().mod_time
-    if dest_mod_time < source_mod_time:
-        if sync_type == 'local-to-b2':
-            yield B2UploadAction(
-                dest_folder.make_full_path(source_file.name), source_file.name, source_mod_time
-            )
-        else:
-            yield B2DownloadAction(source_file.name, source_file.latest_version().id_)
-    if source_mod_time == 0 and dest_mod_time != 0:
-        if sync_type == 'local-to-b2':
-            yield B2DeleteAction(dest_file.name, dest_file.latest_version().id_)
-        else:
-            yield LocalDeleteAction(dest_file.latest_version().id_)
-    # TODO: clean up file history in B2
-    # TODO: do not delete local files for history_days days
 
+    # By default, all but the current version at the destination are
+    # candidates for cleaning.  This will be overridden in the case
+    # where there is no source file.
+    dest_versions_to_clean = []
+    if dest_file is not None:
+        dest_versions_to_clean = dest_file.versions[1:]
 
-def make_folder_sync_actions(source_folder, dest_folder, history_days):
+    # Case 1: Destination does not exist, or source is newer.
+    # All prior versions of the destination file are candidates for
+    # cleaning.
+    transferred = False
+    if dest_mod_time < source_mod_time:
+        yield make_transfer_action(sync_type, source_file, source_folder, dest_folder)
+        transferred = True
+        if sync_type == 'local-to-b2' and dest_file is not None:
+            dest_versions_to_clean = dest_file.versions
+
+    # Case 2: Both exist and source is older
+    elif source_mod_time != 0 and source_mod_time < dest_mod_time:
+        if args.replaceNewer:
+            yield make_transfer_action(sync_type, source_file, source_folder, dest_folder)
+            transferred = True
+        elif args.skipNewer:
+            pass
+        else:
+            raise DestFileNewer('destination file is newer: %s' % (dest_file.name,))
```

Comment:
by our convention this message string should be in `DestFileNewer` exception class and just the filename should be passed in